ambry-commons/src/main/java/com/github/ambry/commons/HostLevelThrottler.java

```diff
@@ -0,0 +1,101 @@
+  private final HardwareUsageMeter hardwareUsageMeter;
+
+  public HostLevelThrottler(HostThrottleConfig hostThrottleConfig) {
+    this.hardwareUsageMeter =
+        new HardwareUsageMeter(hostThrottleConfig.cpuMaxAllowedDelayInMs, hostThrottleConfig.memoryMaxAllowedDelayInMs);
+    JSONObject quota = new JSONObject(hostThrottleConfig.restRequestQuota);
+    JSONObject thresholds = new JSONObject(hostThrottleConfig.hardwareThresholds);
+    quotaMap = new HashMap<>();
+    hardwareThresholdMap = new HashMap<>();
+    for (RestMethod restMethod : RestMethod.values()) {
+      quotaMap.put(restMethod, new RejectThrottler(quota.optInt(restMethod.name(), -1)));
+    }
+    // All criteria should be satisfied.
+    // Future work: if multiple throttling rules under "OR" condition needed, a list of thresholds/hardwareThresholdMap can be used.
+    for (HardwareResource hardwareResource : HardwareResource.values()) {
+      if (thresholds.has(hardwareResource.name())) {
+        JSONObject jsonObject = thresholds.getJSONObject(hardwareResource.name());
+        ObjectMapper objectMapper = new ObjectMapper();
+        try {
+          Criteria criteria = objectMapper.readValue(jsonObject.toString(), Criteria.class);
+          hardwareThresholdMap.put(hardwareResource, criteria);
+        } catch (IOException e) {
+          logger.error("Error while parsing {} threshold ", hardwareResource.name(), e);
+          continue;
+        }
+      }
+    }
+    logger.info("Host throttling config: {} {}", quotaMap, hardwareThresholdMap);
+  }
+
+  public HostLevelThrottler(Map<RestMethod, RejectThrottler> quotaMap,
+      Map<HardwareResource, Criteria> hardwareResourcesCriteriaMap, HardwareUsageMeter hardwareUsageMeter) {
+    this.quotaMap = quotaMap;
+    this.hardwareThresholdMap = new HashMap<>();
+    this.hardwareUsageMeter = hardwareUsageMeter;
+  }
+
+  /**
+   * Return {@code true} if throttling is required. Return {@code false} if no throttler for this rest method or quota
+   * is not reached.
+   * @param restRequest provides the information.
+   */
+  public boolean shouldThrottle(RestRequest restRequest) {
+    boolean shouldThrottle;
+    // Check quotaMap
+    RejectThrottler throttler = quotaMap.get(restRequest.getRestMethod());
+    shouldThrottle = (throttler != null && throttler.shouldThrottle(1));
+    // Check hardwareThresholdMap
+    if (!shouldThrottle && hardwareThresholdMap.containsKey(HardwareResource.MEMORY)) {
+      shouldThrottle = !hardwareThresholdMap.get(HardwareResource.MEMORY)
+          .meetRequirement((long) hardwareUsageMeter.getMemoryPercentage());
+      if (shouldThrottle) {
+        logger.warn("Throttle because of Memory: {}%", hardwareUsageMeter.getMemoryPercentage());
+      }
+    }
+
+    if (!shouldThrottle && hardwareThresholdMap.containsKey(HardwareResource.CPU)) {
+      shouldThrottle = !hardwareThresholdMap.get(HardwareResource.MEMORY)
```

Comment:
do you mean hardwareThresholdMap.get(HardwareResource.CPU)?

Reply:
fixed.